### ISSUE TITLE
Consider Turning ContentProvider Into a Non-Static Class

### ISSUE THREAD

**Issue by ilexp** [AdamsLair/duality#343]
#### Summary

As part of the general maintainability and testability push with Duality, one of the goals is reducing the amount of static API and extracting self-contained instance classes where possible. `ContentProvider` would be an excellent candidate for this.
#### Analysis
- Make `ContentProvider` a non-static class and introduce `IDisposable` API to allow for content cleanups.
- Since static access is still required, have a singleton-like instance in a static `DualityApp` field.
- The new API would likely be something like `DualityApp.Content.Request(xy)`, which isn't much less convenient than before and even comes with some user-side advantages such as extension method support.

**Comment by ilexp** [comment]
## Progress

- Made ContentProvider a non-static class, with a singleton-like global instance stored in `DualityApp.Content`.
- Renamed ContentProvider API to avoid the redundant use of the word `Content` in nearly every method name.
- Resolve handlers now have precedence over the default Load behavior.
- Extracted default content handling code from ContentProvider and moved it into a static internal `DefaultContent` utility class.
- Extracted static, path-based utility methods `GetNameFromPath`, `GetFullNameFromPath` and `IsDefaultContentPath` from ContentProvider and moved them to the Resource class.
- Removed special cases that would prevent users from removing any default content from `ContentProvider`. As non-global instances would now be possible, there should be full control over each instance.

## ToDo

- Write unit tests for `ContentRef<T>`.
- Write unit tests for `ContentProvider`.
- Review the new API and changes and address any remaining issues.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
After the initial work on this, it should be reconsidered whether to follow through on this issue. Turning `ContentProvider` into a non-static class has advantages, but also introduces a bit of a design dissonance: As long as `ContentRef<T>` and other classes are hardwired to use a single, shared instance, resource management remains essentially static, and all other `ContentProvider` instances remain non-functional. Loading a `Texture` via a custom provider will for example trigger loading its dependent `Pixmap`, but do so using the shared instance, not the custom one.

Secondly, it remains an unresolved issue how to handle multiple providers in the first place. A `ContentRef` can only work when knowing where to request resources from, and having multiple places for this complicates the situation a lot with very little gain. Can the same content path now mean different things depending on which provider is asked? Can a ref that failed with provider A recover using provider B? Are refs locked to providers? What happens to a ref using a provider that was disposed?

Overall, it seems like a static, non-instance setup just makes a lot of sense in the `ContentProvider` case. It should be considered to revert all changes made as part of this issue, but retain some of the changes that are unrelated to the static / non-static change, such as:

- Resolve / Load handler precedence.
- Extracting default content handling code, to be moved to a static `DefaultContent` utility class or similar.
- Extracting path-based utility methods, to be moved to `Resource`.
- Removing default content special cases, so default content can be removed just the same as any other content.
- Consider keeping some of the API changes that were introduced.

**Comment by ilexp** [comment]
## Progress

- Turned `ContentProvider` back into a static class.
- Renamed most `ContentProvider` API back into including the "content" keyword, as it's no longer redundant with the static API.
- Added unit tests for basic `ContentProvider` functionality.
- Fixed a bug in `RenameContentTree` for paths including repetitions.
- Improved `ContentRef` recovery to work based on disposed `Resource` instances by keeping their path property valid on disposal.

## ToDo

- Write more unit tests for `ContentRef<T>` and `ContentProvider`.
- Squash-Merge back to `develop-3.0`.

**Comment by ilexp** [comment]
## Progress

- Added more unit tests for `ContentProvider`. Skipping `ContentRef` tests for now, as they're non-breaking and non-critical for now.
- Squash-Merged back to `develop-3.0`

Didn't manage for some reason to include more than one line into the squashed commit message using SourceTree. Is this a bug? Not sure. In any case, here's the missing commit message part / changelog:

```
ContentProvider Bugfixes, Tweaks, API Changes, Tests

#ADD: Added unit tests for the most basic functionality of the ContentProvider.
#FIX: Fixed a bug in ContentProvider renaming behavior where Resource paths with repetitions would fail to be renamed correctly.
#CHANGE: Resources now retain their original path when disposed, allowing ContentRefs to recover based on disposed resources.
#CHANGE: Removing content from the ContentProvider now handles cases where a single Resource is registered on multiple path keys.
#CHANGE: Added content to the ContentProvider on an already used path will now dispose the previously registered Resource.
#CHANGE: Moved specialized code for discarding plugin related data out of ContentProvider and directly into the DualityApp event handler that requires it.
#CHANGE: Moved static, resource path based API from ContentProvider to Resource.
#CHANGE: Moved loading / handling of embedded default content from ContentProvider and Resource to internal static DefaultContent utility class.
#CHANGE: Default content is no longer treated with a readonly special case in ContentProvider, and a call to Clear will remove both default and non-default content.
#CHANGE: Renamed ContentRef.MakeAvailable to EnsureLoaded
#CHANGE: Internal handling of default content no longer requires a specific init code segment, and instead tracks additions and removals.
#CHANGE: Resource Resolve handlers now take precedence over automatically loading Resources from the Data directory.
```
